### PR TITLE
fix(field): store/sync a value instead of delegating

### PR DESCRIPTION
**Pull request by daKmoR**

### CODE REVIEW

**Comment by tlouisse** [comment]
This is about the modelValue and value being out of sync, right? (or more precise, LionInput.value and LionInput.$$slot('input').value being out of sync).

Before we do a proper proposal for a rewrite of DelegateMixin, maybe we can do a quick fix like this (pseudo code):

```js
// LionField...

set value(v) {
   this.__value = v;
   // sync with $$slot('input') when available
}

get value() {
   return this.__value;
}
connectedCallback() {
   // for all cases to work, we need to make sure we add logic before the SlotMixin does its thing:
   // Now, listen for slotChanged event and sync the 'cached' this.value to $$slot('input');
   super.connectedCallback();
}
```

A proper fix makes sure delegated values that are in 'external' dom not managed by LionField (in this case the light dom created by SlotMixin are synced whenever  external dom is available for the first time.

**Comment by daKmoR** [comment]
> This is about the modelValue and value being out of sync, right?

yes

> (or more precise, LionInput.value and LionInput.$$slot('input').value being out of sync).

no

e.g. it's about that you set a `modelValue` at one time and the `lion-input` actually has that `modelValue` but it's not visible to the user as the `value` (both for `lion-input` and `input`) is still empty

**Comment by tlouisse** [comment]
> yes
> 
> > (or more precise, LionInput.value and LionInput.$$slot('input').value being out of sync).
> 
> no
> 
> e.g. it's about that you set a `modelValue` at one time and the `lion-input` actually has that `modelValue` but it's not visible to the user as the `value` (both for `lion-input` and `input`) is still empty

True, but if I remember right `modelValue -> reflectbackFormattedValueToUser (sets this.value) -> this.value will sync to $$slot('input').value`.
Only the last step will go wrong I guess

**Comment by bashmish** [comment]
@daKmoR is this related to https://github.com/ing-bank/lion/issues/22 ?

**Comment by daKmoR** [comment]
I guess these are both side effects of the underlying problem of syncing of `lion-input.value` and `nativeInput.value`

**Comment by tlouisse** [comment]
https://github.com/ing-bank/lion/issues/22 is purely about the order of application of mixins. Because validateMixin is applied after FormatMixin, the observed `_modelValueChanged` responds later ->
as a consequence we don't have a correct errorState at the time we want to format.

Seems unrelated to the syncing of the `value` from LionInput to native `<input>`

**Comment by tlouisse** [comment]
About the failing test: it accidentally instantiates `ing-input` instead of `lion-input`. If I change that, the test just passes.
@ruphin do you still have this issue?

**Comment by tlouisse** [comment]
Update: spoke to @ruphin and the bug seems to be the other way around: 
setting `.value` programmatically is not synced to `.modelValue`.

This pr handles:
- fix the test
- move all value 'loop' logic to FormatMixin (it previously was on LionField)
- make value a prop so that it will maintain state on the host instead of on the child




Still TODO: make nice commit history....

File: packages/fieldset/test/lion-fieldset.test.js [hunk 1]
```diff
@@ -55,16 +70,19 @@ describe('<lion-fieldset>', () => {
...
     await nextFrame();
     fieldset.formElements['gender[]'][0].modelValue = { value: 'male' };
     fieldset.formElements['hobbies[]'][0].modelValue = { checked: false, value: 'chess' };
     fieldset.formElements['hobbies[]'][1].modelValue = { checked: false, value: 'rugby' };
 
     expect(Object.keys(fieldset.formElements).length).to.equal(3);
     expect(fieldset.formElements['hobbies[]'].length).to.equal(2);
+
     expect(fieldset.formElements['hobbies[]'][0].modelValue.value).to.equal('chess');
-    expect(fieldset.formElements['gender[]'][0].modelValue.value).to.equal('male');
```
Comment (bashmish) [inline packages/fieldset/test/lion-fieldset.test.js]: why is this removed?

File: packages/fieldset/test/lion-fieldset.test.js [hunk 2]
```diff
@@ -9,19 +9,34 @@ import {
...
+  class extends LionInput {
+    parser() {
+      return this.modelValue;
+    }
+
+    formatter(mv) {
+      return mv && mv.value !== undefined ? mv.value : mv;
+    }
+  },
+);
```
Comment (bashmish) [inline packages/fieldset/test/lion-fieldset.test.js]: please use `defineCE` helper from open-wc to prevent name collisions in our tests

remember that we run all tests (more than 700 now) on the same page without iframes?

File: packages/field/test/FormatMixin.test.js [hunk 3]
```diff
@@ -124,7 +124,12 @@ describe('FormatMixin', () => {
     expect(fooFormat.modelValue).to.equal('string');
   });
 
-  it('synchronizes inputElement.value as a fallback mechanism', async () => {
+  it('converts value to modelValue (for platform api comtatibility)', async () => {
```
Comment (bashmish) [inline packages/field/test/FormatMixin.test.js]: `comtatibility` => `compatibility`

File: packages/field/src/LionField.js [hunk 4]
```diff
@@ -136,60 +123,24 @@ export class LionField extends FormControlMixin(
...
-    Event Handlers */
-
+  // Needed for choice-inputs and select (IE11) as an alternative to the input event.
+  // TODO: add in FormatMixin and remove here and in lion-select
   _onChange() {
     if (super._onChange) super._onChange();
     this.dispatchEvent(
       new CustomEvent('user-input-changed', {
         bubbles: true,
+        composed: true,
```
Comment (bashmish) [inline packages/field/src/LionField.js]: previously this bubbled, meaning that it will reach all parent elements in the same shadow root where the host (custom input) is located, so for forms and custom forms inputs which are entirely located in the same shadow root (by our convention) it was enough

what changed? do we need some elements extending from LionField to be in a deeper shadow root than the rest of the form? can you show the code you are targeting, like a DOM structure with all elements and shadow roots, roughly?

Comment (tlouisse) [reply packages/field/src/LionField.js]: We fire same event in different places (with composed), there is no reason to do it without in this particular case. Or more generic, there is no reason to be inconsistent throughout our code

File: packages/fieldset/test/lion-fieldset.test.js (hunk 2 above)
Comment (tlouisse) [reply packages/fieldset/test/lion-fieldset.test.js]: Chances are slim and readability will go down, but will do ;)

File: packages/field/src/FormatMixin.js [hunk 5]
```diff
@@ -226,6 +237,49 @@ export const FormatMixin = dedupeMixin(
         this._calculateValues({ source: 'serialized' });
       }
 
+      _onValueChanged({ value }) {
+        if (super._onValueChanged) super._onValueChanged();
```
Comment (jorenbroekema) [inline packages/field/src/FormatMixin.js]: Use brackets { 

}

Note: should probably set a linting rule to make it consistent, in Poland it was pointed out by someone that we are inconsistent on this.

Comment (bashmish) [reply packages/field/src/FormatMixin.js]: fully agree with having a linter rule active for this

I saw too many times how this ended up with bugs eventually

Comment (bashmish) [reply packages/field/src/FormatMixin.js]: I think we'll work on this later, there is a lot to change here and this MR is not about removing all the technical debt, but rather fixing one specific issue with the logic, so most of the code is just moved from one file to another

File: packages/fieldset/test/lion-fieldset.test.js [hunk 6]
```diff
@@ -6,22 +6,38 @@ import {
...
 import '../lion-fieldset.js';
 
+const complexInputTagString = defineCE(
+  class extends LionInput {
+    parser() {
+      return this.modelValue;
+    }
+
+    formatter(mv) {
+      return mv && mv.value !== undefined ? mv.value : mv;
```
Comment (bashmish) [inline packages/fieldset/test/lion-fieldset.test.js]: this makes all text inputs in tests to behave like special case inputs, and according to @tlouisse this was implemented to fix some tests that contained text inputs with complex model values (❗️) and started failing after the fix

I think this is a symptom of wrong tests in the first place and those wrong tests should have been fixed instead of introducing this special cases for every single test in this file

also I think we have pretty bad practice in this test file where we reuse a template (`inputSlot` defined below) in every test while not every element defined in it is tested, I'd rather have a needed templated defined in each test separately and make tests more isolated, also preventing other issue like the one described earlier in this comment

File: packages/field/src/FormatMixin.js [hunk 7]
```diff
@@ -300,6 +358,18 @@ export const FormatMixin = dedupeMixin(
...
       }
 
+      // Needed for choice-inputs and select (IE11) as an alternative to the input event.
+      // TODO: add in FormatMixin and remove here and in lion-select
+      _onChange() {
+        if (super._onChange) super._onChange();
+        this.dispatchEvent(
+          new CustomEvent('user-input-changed', {
+            bubbles: true,
+            composed: true,
```
Comment (bashmish) [inline packages/field/src/FormatMixin.js]: you didn't only copy, but also added `composed: true`

this is not needed, and adding it for some consistency is a bad idea, because you create an extra event which will bubble outside of shadow root while this is not expected by developers... so this fix to consistency creates additional problems without the real need for it

for now I would just leave it as it was before (without `composed: true`) and in the long run revisit all events we dispatch and double check if they need to bubble and bubble outside of shadow roots or not, and the decision in each case will be individual based on our API needs and users needs

File: packages/field/src/LionField.js (hunk 4 above)
Comment (tlouisse) [reply packages/field/src/LionField.js]: To simplifiy this merge request, I will put this back in original file (and state). In a later refactor, we can move this logic to one single, appropriate place with shared functionality for ChoiceInput and Select

File: packages/field/src/FormatMixin.js (hunk 5 above)
Comment (tlouisse) [reply packages/field/src/FormatMixin.js]: Agreed. I would agree with a linting rule for this

File: packages/fieldset/test/lion-fieldset.test.js (hunk 6 above)
Comment (tlouisse) [reply packages/fieldset/test/lion-fieldset.test.js]: Fair point 👍  I will look at this later again, when I have an overview of the impact of all pending form core updates

File: packages/field/src/FormatMixin.js (hunk 7 above)
Comment (tlouisse) [reply packages/field/src/FormatMixin.js]: Fair point again.
I will do an attempt to document everything Event related:
- no CustomEvents
- bubble only when needed
- composed only when needed

Comment (bashmish) [reply packages/field/src/FormatMixin.js]: @tlouisse ha-ha, add one more :D https://twitter.com/JonathanAschan/status/1138744064835280897

**Comment by daKmoR** [comment]
ok this needs a better approach... might be superseded by https://github.com/ing-bank/lion/pull/260/commits
